src/syntax/rules.rkt

```diff
@@ -68,13 +44,26 @@
       (match-define (list rules groups types) ruleset)
       (list (filter rule-ops-supported? rules) groups types)))))
 
+;; Note on rules
+;; fp-safe-simplify ⊂ simplify ⊂ all
+;;
+;; all                    requires at least one tag of an active group of rules
+;; simplify               same req. as all + 'simplify' tag
+;; fp-safe-simplify       same req. as simplify + 'fp-safe' tag ('fp-safe' does not imply 'simplify')
+;;
+
 (define (update-rules rules groups)
-  (when (ormap (curry flag-set? 'rules) groups)       ; update all
-    (all-rules (append (all-rules) rules))
-    (when (set-member? groups 'simplify)              ; update simplify
-      (simplify-rules (append (simplify-rules) rules))  
-      (when (set-member? groups 'fp-safe)         ; update fp-safe
-        (fp-safe-simplify-rules (append (fp-safe-simplify-rules) rules))))))
+  (when (ormap (curry flag-set? 'rules) groups)             ; update all
+    (*rules* (append (*rules*) rules))
+    (when (set-member? groups 'simplify)                    ; update simplify
+      (*simplify-rules* (append (*simplify-rules*) rules))
+      (when (set-member? groups 'fp-safe)                   ; update fp-safe
+        (*fp-safe-simplify-rules*
+          (append (*fp-safe-simplify-rules*) rules))))))
```

Comment:
Do we think caching the rule types is actually performance-critical? Having a cache & updating it is definitely more error-prone than just recomputing each time, and IIRC we only call the `*rules*` functions a couple of times per run.

Reply:
I think expanding the rules each time is marginally cleaner now that all egg-based rewriting goes through the same interface. Perhaps it can be cached there, but in general I don't think handling rules takes very long.

Reply:
Right, what I'm saying is is that right now we store the `*simplify-rules*` and `*fp-safe-simplify-rules*` as separate lists, but perhaps we should just have one list of all the rules, and then those two subsets of rules are just functions that filter the list of rules every time they're called.

Reply:
Ahh interesting. Back to the old method. Might be okay. I'll just leave it as is for now.